### PR TITLE
Add Phi3

### PR DESCRIPTION
**Pull request by nailimixaM**
Added tests: `pytest -m quarot` for unit tests, `pytest -m quarot_experiment` for longer e2e tests. All require gpu.

Includes changes of #154 and #152.

Co-authored with @pashminacameron @jameshensman.

### CODE REVIEW

File: experiments/run_quarot.py [hunk 1]
```diff
@@ -276,21 +274,32 @@ def quarot_main(args: argparse.Namespace) -> None:
             dataset=dataset["train"], tokenizer=tokenizer, batch_size=args.ppl_eval_batch_size
         )
 
-        quarot_model_adapter = LlamaModelAdapter(quarot_llama)
+        if isinstance(quarot_model, QuarotLlamaForCausalLM):
```
Comment (jameshensman) [inline experiments/run_quarot.py:300]: is the long-term design here to have if-else switches for each model? Or is there a cleaner way to match up model adapters?

File: src/quarot/gptq.py [hunk 2]
```diff
@@ -23,23 +23,36 @@ def quantize_weight_gptq(
...
+
+    Args:
+        W: The weight tensor to quantize.
+        H: The Hessian of the activations, must have dtype==torch.float32.
+        bits: The number of bits to quantize to.
+        symmetric: Whether to use symmetric quantization.
+        max_blocksize: The maximum block size for the GPTQ algorithm.
+        percdamp: The damping factor for the Hessian.
+        groupsize: The number of weights to group together for quantization.
+        clip_weights: Whether to clip the weights (by searching for good clipping ratios) to the range of the quantization.
```
Comment (jameshensman) [inline src/quarot/gptq.py:63]: Can we add to the docstring some infor on what gets returned, and shat dtype/shape each is. Perhaps

```
Returns:
    Q: integer quantized weights, shape (out_features, in_features)
    scales: floating-point scales for the quantization, shape (out_features, num_groups)
    offsets: offsets for quantization. shape (out_features, num_groups). None if using symmetric quantization. 
```

File: src/quarot/gptq.py [hunk 3]
```diff
@@ -23,23 +23,36 @@ def quantize_weight_gptq(
...
+    else:
+        raise NotImplementedError("Groupwise quantization not yet tested for GPTQ")
 
     # find dead weights and set them to zero, and their Hess value to 1
-    dead = torch.diag(H) == 0
-    H[dead, dead] = 1
-    W[:, dead] = 0
+    dead = torch.diag(H) == 0.0
+    H[dead, dead] = 1.0
+    W[:, dead] = 0.0
```
Comment (jameshensman) [inline src/quarot/gptq.py:87]: perhaps we could log the number of dead weights? it should surely be zero for any modern model!

File: src/quarot/gptq.py [hunk 4]
```diff
@@ -213,7 +221,9 @@ def quantize_model_gptq(
     apply_mask: bool = False,
 ) -> None:
     """
-    TODO
+    Quantize the model in-place using the GPTQ scheme, using the dataloader calibration data. All weights are stored in FP16.
+    If the model is a QuaRot model, the weights-minus-offsets and scales are stored in the QuarotFP16Linear modules. If the model is not a QuaRot model,
```
Comment (jameshensman) [inline src/quarot/gptq.py:296]: wouldn't it be cleaner for the QuaRotLinear module to contain an optional offset?

**Review by jameshensman — APPROVED** [review]
holy schmoly the main GPTQ loop is now so clean. Love the patterns in tests too.

File: experiments/run_quarot.py (hunk 1 above)
Comment (pashminacameron) [reply experiments/run_quarot.py:300]: This is on me. Read 'paper deadline coding' :) We should be able to clean this up like slicegpt adapters even though these are huggingface classes by using class name association like we do for compressedlayers and also move it out of here into `hf_utils` or like.

**Review by pashminacameron — APPROVED** [review]

**Pull request merged**